cmd/storer2elastic/dataprocessor/dataProcessor.go

```diff
@@ -0,0 +1,514 @@
+	if err != nil {
+		return nil, err
+	}
+	metaPersHolder.hdrHashNoncePersister = headerHashNonceUnit
+
+	return metaPersHolder, nil
+}
+
+func (dp *dataProcessor) closeMetaPersisters(persistersHolder *metaBlocksPersistersHolder) {
+	err := persistersHolder.metaBlocksPersister.Close()
+	log.LogIfError(err)
+
+	err = persistersHolder.hdrHashNoncePersister.Close()
+	log.LogIfError(err)
+}
+
+func (dp *dataProcessor) getEpochStartMetaBlock(record *databasereader.DatabaseInfo, metaPersisters *metaBlocksPersistersHolder) (*block.MetaBlock, error) {
+	epochStartIdentifier := core.EpochStartIdentifier(record.Epoch)
+	metaBlockBytes, err := metaPersisters.metaBlocksPersister.Get([]byte(epochStartIdentifier))
+	if err == nil && len(metaBlockBytes) > 0 {
+		return dp.unmarshalMetaBlock(metaBlockBytes)
+	}
+	log.Warn("epoch start meta block not found", "epoch", record.Epoch)
+
+	// header should be found by the epoch start identifier. if not, try getting the nonce 0 (the genesis block)
+	return dp.getMetaBlockForNonce(0, record, metaPersisters)
+}
+
+func (dp *dataProcessor) getMetaBlockForNonce(nonce uint64, record *databasereader.DatabaseInfo, metaPersisters *metaBlocksPersistersHolder) (*block.MetaBlock, error) {
+	nonceBytes := dp.uint64Converter.ToByteSlice(nonce)
+	metaBlockHash, err := metaPersisters.hdrHashNoncePersister.Get(nonceBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	metaBlockBytes, err := metaPersisters.metaBlocksPersister.Get(metaBlockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return dp.unmarshalMetaBlock(metaBlockBytes)
+}
+
+func (dp *dataProcessor) processMetaBlock(
+	metaBlock *block.MetaBlock,
+	dbsInfo []*databasereader.DatabaseInfo,
+	record *databasereader.DatabaseInfo,
+	metaBlocksPersisters *metaBlocksPersistersHolder,
+	persisters *persistersHolder,
+	shardPersisters map[uint32]*persistersHolder,
+) error {
+	err := dp.indexHeader(persisters, record, metaBlock, record.Epoch)
+	if err != nil {
+		return err
+	}
+
+	for _, shardInfo := range metaBlock.ShardInfo {
+		err = dp.processShardInfo(dbsInfo, &shardInfo, metaBlock.Epoch, shardPersisters[shardInfo.ShardID])
+		log.LogIfError(err)
+	}
+
+	return nil
+}
+
+func (dp *dataProcessor) processShardInfo(
+	dbsInfos []*databasereader.DatabaseInfo,
+	shardInfo *block.ShardData,
+	epoch uint32,
+	shardPersisters *persistersHolder,
+) error {
+	shardDBInfo, err := getShardDatabaseForEpoch(dbsInfos, epoch, shardInfo.ShardID)
+	if err != nil {
+		return err
+	}
+
+	shardHeader, err := dp.getShardHeader(shardPersisters, shardInfo.HeaderHash)
+	if err != nil {
+		// if new epoch, shard headers can be found in the previous epoch's storage
+		// TODO: for this case return and use all the persisters from the previous epoch in order to find txs
+		shardHeader, err = dp.getFromShardStorer(dbsInfos, shardInfo, epoch-1)
+		if err != nil {
+			return err
+		}
+	}
+
+	return dp.indexHeader(shardPersisters, shardDBInfo, shardHeader, epoch)
```

Comment:
This can be another processor that just index headers.

Reply:
created a separate component which reads from the dataReplayer and indexes data